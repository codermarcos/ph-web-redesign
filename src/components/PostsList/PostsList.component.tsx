import { useEffect } from 'react';
import { useInView } from 'react-intersection-observer';

import { List } from '@/components/PostsList/PostsList.styles';

import PostListItem, {
  ListItem,
} from '@/components/PostsListItem/PostsListItem.component';

type Props = {
  edges: Array<{ node: ListItem }>;
  onLoadMore: () => void;
}

function PostsList({ edges, onLoadMore }: Props) {

  const { ref, inView } = useInView({ threshold: 0.1 });

  useEffect(
    () => {

      if (inView)
        onLoadMore();

    },
    [
      inView,
      onLoadMore,
    ]
  );

  return (
    <List>
      {
        edges.map(
          ({ node }) => (
            <PostListItem
              key={node.id}
              {...node}
            />
          )
        )
      }

      <li ref={ref}>
        loading more...
      </li>
    </List>
  );

}

export default PostsList;
