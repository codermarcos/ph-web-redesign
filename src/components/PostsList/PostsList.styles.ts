import styled from 'styled-components';

import { px } from 'csx';

export const List = styled.ol({
  listStyle: 'none',
  marginRight: px(20),
});
