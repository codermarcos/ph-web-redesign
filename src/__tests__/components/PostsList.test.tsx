import { RenderResult, render } from '@testing-library/react';

import { useInView } from 'react-intersection-observer';

import { faker } from '@faker-js/faker';

import type { ListItem } from '@/components/PostsListItem/PostsListItem.component';

import PostsList from '@/components/PostsList/PostsList.component';

jest.mock(
  'react-intersection-observer',
  () => ({
    useInView: jest.fn(),
  })
);

faker.seed(1);

const mockPosts: Array<{ node: ListItem }> = [
  {
    node: {
      id: faker.random.alphaNumeric(5),
      name: faker.name.fullName(),
      tagline: faker.lorem.words(8),
      thumbnail: {
        url: faker.image.abstract(),
      },
      votesCount: parseInt(faker.random.numeric(2), 10),
    },
  },
];

describe(
  '[Component] -> components/PostsList',
  () => {

    // eslint-disable-next-line init-declarations
    let wrapper: RenderResult;
    // eslint-disable-next-line init-declarations
    let onLoadSpy: jest.Mock;

    beforeEach(
      () => {

        onLoadSpy = jest.fn();

        (useInView as jest.Mock).mockReturnValue({});

        wrapper = render(
          <PostsList
            edges={mockPosts}
            onLoadMore={onLoadSpy}
          />
        );

      }
    );

    it(
      'should render correct link',
      () => {

        expect(wrapper.container).toMatchSnapshot();

      }
    );

  }
);

