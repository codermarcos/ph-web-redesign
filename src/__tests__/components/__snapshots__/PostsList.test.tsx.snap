// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] -> components/PostsList should render correct link 1`] = `
<div>
  <ol
    class="sc-af6e6f56-0 hReOgG"
  >
    <li
      class="sc-8132b4de-0 kJObhu"
    >
      <a
        class="sc-8132b4de-4 egGXMv"
        href=""
      >
        <figure
          class="sc-8132b4de-1 jrYBBL"
        >
          <img
            alt="Image for Alicia Zulauf"
            class="sc-8132b4de-2 dgPkgN"
            data-nimg="1"
            decoding="async"
            height="96"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Floremflickr.com%2F640%2F480%2Fabstract&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Floremflickr.com%2F640%2F480%2Fabstract&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Floremflickr.com%2F640%2F480%2Fabstract&w=256&q=75 2x"
            style="color: transparent;"
            width="96"
          />
          <figcaption
            class="sc-8132b4de-3 eZTIlJ"
          >
            <b>
              Alicia Zulauf
            </b>
            <p>
              et aliquid qui laboriosam nostrum est aliquid earum
            </p>
          </figcaption>
        </figure>
      </a>
      <button
        class="sc-8132b4de-5 cONJLt"
        type="button"
      >
        58
      </button>
    </li>
    <li>
      loading more...
    </li>
  </ol>
</div>
`;
